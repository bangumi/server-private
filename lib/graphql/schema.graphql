type Query {
    me: User
    character(id: Int!): Character
    person(id: Int!): Person
    subject(id: Int!): Subject
}

type Avatar {
    large: String!
    medium: String!
    small: String!
}

"""basic character info as field of other types to avoid recursive query"""
type SlimCharacter {
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    redirect: Int!
    role: Int!
    summary: String!
}

type Character {
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    persons(limit: Int! = 10, offset: Int! = 0): [CharacterRelatedPerson!]
    redirect: Int!
    role: Int!
    subjects(limit: Int! = 10, offset: Int! = 0): [CharacterRelatedSubject!]
    summary: String!
}

type CharacterRelatedPerson {
    person: SlimPerson!
    subject: SlimSubject!
    summary: String!
}

type CharacterRelatedSubject {
    order: Int!
    subject: SlimSubject!
    type: Int!
}

type Episode {
    airdate: String!
    comment: Int!
    description: String!
    disc: Int!
    duration: String!
    id: Int!
    last_post: Int!
    name: String!
    name_cn: String!
    sort: Float!
    type: Int!
}

type Images {
    grid: String!
    large: String!
    medium: String!
    small: String!
}

type Infobox {
    key: String!
    values: [InfoboxValue!]
}

type InfoboxValue {
    k: String
    v: String
}

"""basic person info as field of other types to avoid recursive query"""
type SlimPerson {
    career: [String!]!
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    redirect: Int!
    summary: String!
    type: Int!
}

type Person {
    career: [String!]!
    characters(limit: Int! = 10, offset: Int! = 0): [PersonRelatedCharacter!]
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    redirect: Int!
    subjects(limit: Int! = 10, offset: Int! = 0): [PersonRelatedSubject!]
    summary: String!
    type: Int!
}

type PersonRelatedCharacter {
    character: SlimCharacter!
    subject: SlimSubject!
    summary: String!
}

type PersonRelatedSubject {
    position: Int!
    subject: SlimSubject!
}

"""a subject as field of other types to avoid recursive query"""
type SlimSubject {
    airtime: SubjectAirtime!
    eps: Int!
    id: Int!
    images: SubjectImages
    infobox: [Infobox!]
    locked: Boolean!
    name: String!
    name_cn: String!
    nsfw: Boolean!
    platform: SubjectPlatform!
    rating: SubjectRating!
    redirect: Int!
    series: Boolean!
    series_entry: Int!
    summary: String!
    type: Int!
    volumes: Int!
}

type Subject {
    airtime: SubjectAirtime!
    characters(limit: Int! = 10, offset: Int! = 0): [SubjectRelatedCharacter!]
    collection: SubjectCollection!
    episodes(
        limit: Int! = 30

        """超出范围时会返回空数据。可以使用负数，来获取最后几个章节。如 `-1` 会返回最后一个章节， `-2` 会返回最后两个章节"""
        offset: Int! = 0
        type: Int
    ): [Episode!]
    eps: Int!
    id: Int!
    images: SubjectImages
    infobox: [Infobox!]
    locked: Boolean!
    name: String!
    name_cn: String!
    nsfw: Boolean!
    persons(limit: Int! = 10, offset: Int! = 0): [SubjectRelatedPerson!]
    platform: SubjectPlatform!
    rating: SubjectRating!
    redirect: Int!
    relations(excludeTypes: [Int!], includeTypes: [Int!], limit: Int! = 30, offset: Int! = 0): [SubjectRelation!]
    series: Boolean!
    series_entry: Int!
    summary: String!
    tags(limit: Int = 0): [SubjectTag!]!
    topics(limit: Int! = 10, offset: Int! = 0): [SubjectTopic!]
    type: Int!
    volumes: Int!
}

type SubjectAirtime {
    date: String!
    month: Int!
    weekday: Int!
    year: Int!
}

type SubjectCollection {
    collect: Int!
    doing: Int!
    dropped: Int!
    on_hold: Int!
    wish: Int!
}

type SubjectImages {
    common: String!
    grid: String!
    large: String!
    medium: String!
    small: String!
}

type SubjectPlatform {
    alias: String
    id: Int!
    type: String!
    type_cn: String
}

type SubjectRating {
    count: [Int!]
    rank: Int!
    score: Float!
    total: Int!
}

type SubjectRelatedCharacter {
    character: SlimCharacter!
    order: Int!
    type: Int!
}

type SubjectRelatedPerson {
    person: SlimPerson!
    position: Int!
}

type SubjectRelation {
    order: Int!
    relation: Int!
    subject: SlimSubject!
}

type SubjectTag {
    count: Int!
    name: String!
}

type SubjectTopic {
    created_at: Int!
    creator: User!
    display: Int!
    id: Int!
    replies: Int!
    state: Int!
    title: String!
    updated_at: Int!
}

type User {
    avatar: Avatar!
    id: Int!
    nickname: String!
    username: String!
}
