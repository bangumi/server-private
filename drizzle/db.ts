import type { ExtractTablesWithRelations } from 'drizzle-orm';
import type { MySqlTransaction } from 'drizzle-orm/mysql-core';
import {
  drizzle,
  type MySql2PreparedQueryHKT,
  type MySql2QueryResultHKT,
} from 'drizzle-orm/mysql2';
import * as mysql from 'mysql2/promise';

import config, { developing, testing } from '@app/lib/config.ts';
import { logger } from '@app/lib/logger.ts';

import * as schema from './schema.ts';

const poolConnection = mysql.createPool({
  host: config.mysql.host,
  user: config.mysql.user,
  database: config.mysql.db,
  password: config.mysql.password,
  port: config.mysql.port,
});

export const db = drizzle(poolConnection, {
  schema,
  mode: 'default',
  logger:
    testing || developing
      ? {
          logQuery(query: string, params: unknown[]) {
            logger.trace('query', { query, params });
          },
        }
      : undefined,
});

export * as op from 'drizzle-orm';

export type Txn = MySqlTransaction<
  MySql2QueryResultHKT,
  MySql2PreparedQueryHKT,
  typeof schema,
  ExtractTablesWithRelations<Record<string, unknown>>
>;
