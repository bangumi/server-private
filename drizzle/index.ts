export * from './db.ts';
export * as orm from './orm.ts';
export * as schema from './schema.ts';
export * as op from 'drizzle-orm';
