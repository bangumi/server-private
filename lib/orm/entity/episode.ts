import { Column, Entity, Index, PrimaryGeneratedColumn } from 'typeorm';

@Index('ep_sort', ['sort'], {})
@Index('ep_disc', ['epDisc'], {})
@Index('ep_subject_id', ['subjectID'], {})
@Index('ep_lastpost', ['epLastPost'], {})
@Index('ep_ban', ['epBan'], {})
@Index('ep_subject_id_2', ['subjectID', 'epBan', 'sort'], {})
@Entity('chii_episodes', { schema: 'bangumi' })
export class Episode {
  @PrimaryGeneratedColumn({ type: 'mediumint', name: 'ep_id', unsigned: true })
  id!: number;

  @Column('mediumint', { name: 'ep_subject_id', unsigned: true })
  subjectID!: number;

  @Column('float', {
    name: 'ep_sort',
    unsigned: true,
    precision: 12,
    default: () => "'0'",
  })
  sort!: number;

  @Column('tinyint', { name: 'ep_type', unsigned: true })
  type!: number;

  @Column('tinyint', {
    name: 'ep_disc',
    comment: '碟片数',
    unsigned: true,
    default: () => "'0'",
  })
  epDisc!: number;

  @Column('varchar', { name: 'ep_name', length: 80 })
  name!: string;

  @Column('varchar', { name: 'ep_name_cn', length: 80 })
  nameCN!: string;

  @Column('tinyint', { name: 'ep_rate' })
  epRate!: number;

  @Column('varchar', { name: 'ep_duration', length: 80 })
  duration!: string;

  @Column('varchar', { name: 'ep_airdate', length: 80 })
  airDate!: string;

  /** @deprecated 在线播放地址 */
  @Column('mediumtext', { name: 'ep_online' })
  epOnline!: string;

  @Column('mediumint', { name: 'ep_comment', unsigned: true })
  epComment!: number;

  @Column('mediumint', { name: 'ep_resources', unsigned: true })
  epResources!: number;

  @Column('mediumtext', { name: 'ep_desc' })
  summary!: string;

  @Column('int', { name: 'ep_dateline', unsigned: true })
  epDateline!: number;

  @Column('int', { name: 'ep_lastpost', unsigned: true })
  epLastPost!: number;

  @Column('tinyint', { name: 'ep_lock', unsigned: true, default: () => "'0'" })
  epLock!: number;

  @Column('tinyint', { name: 'ep_ban', unsigned: true, default: () => "'0'" })
  epBan!: number;
}

@Index('ep_cmt_crt_id', ['epID'], {})
@Index('ep_pst_related', ['relatedID'], {})
@Index('ep_pst_uid', ['creatorID'], {})
@Entity('chii_ep_comments', { schema: 'bangumi' })
export class EpisodeComment {
  @PrimaryGeneratedColumn({
    type: 'mediumint',
    name: 'ep_pst_id',
    unsigned: true,
  })
  id!: number;

  @Column('mediumint', { name: 'ep_pst_mid', unsigned: true })
  epID!: number;

  @Column('mediumint', { name: 'ep_pst_uid', unsigned: true })
  creatorID!: number;

  @Column('mediumint', {
    name: 'ep_pst_related',
    unsigned: true,
    default: () => "'0'",
  })
  relatedID!: number;

  @Column('int', { name: 'ep_pst_dateline', unsigned: true })
  createdAt!: number;

  @Column('mediumtext', { name: 'ep_pst_content' })
  content!: string;

  @Column('mediumtext', { name: 'ep_pst_state' })
  state!: number;
}
