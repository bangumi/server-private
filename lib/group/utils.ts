import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
