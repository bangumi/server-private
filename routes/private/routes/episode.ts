import { Type as t } from '@sinclair/typebox';
import { DateTime } from 'luxon';

import { db, op } from '@app/drizzle/db.ts';
import * as schema from '@app/drizzle/schema';
import { NotAllowedError } from '@app/lib/auth/index.ts';
import { Dam } from '@app/lib/dam.ts';
import { BadRequestError, CaptchaError, NotFoundError } from '@app/lib/error.ts';
import { Security, Tag } from '@app/lib/openapi/index.ts';
import { turnstile } from '@app/lib/services/turnstile.ts';
import { CommentState } from '@app/lib/topic/type.ts';
import * as convert from '@app/lib/types/convert.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import * as req from '@app/lib/types/req.ts';
import * as res from '@app/lib/types/res.ts';
import { LimitAction } from '@app/lib/utils/rate-limit';
import { requireLogin } from '@app/routes/hooks/pre-handler.ts';
import { rateLimit } from '@app/routes/hooks/rate-limit';
import type { App } from '@app/routes/type.ts';

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.get(
    '/subjects/-/episodes/:episodeID',
    {
      schema: {
        summary: '获取剧集信息',
        tags: [Tag.Episode],
        operationId: 'getSubjectEpisode',
        params: t.Object({
          episodeID: t.Integer({ examples: [1075440] }),
        }),
        response: {
          200: res.Ref(res.Episode),
        },
      },
    },
    async ({ params: { episodeID } }): Promise<res.IEpisode> => {
      const ep = await fetcher.fetchEpisodeByID(episodeID);
      if (!ep) {
        throw new NotFoundError(`episode ${episodeID}`);
      }
      return ep;
    },
  );

  app.get('/subjects/-/episode/:episodeID', (req, reply) => {
    const params = req.params as Record<string, string>;
    const episodeID = params.episodeID ?? '';
    return reply.redirect(`/p1/subjects/-/episodes/${episodeID}`, 307);
  });

  app.get(
    '/subjects/-/episodes/:episodeID/comments',
    {
      schema: {
        summary: '获取条目的剧集吐槽箱',
        tags: [Tag.Episode],
        operationId: 'getSubjectEpisodeComments',
        params: t.Object({
          episodeID: t.Integer({ examples: [1075440], minimum: 0 }),
        }),
        response: {
          200: t.Array(res.EpisodeComment),
        },
      },
    },
    async ({ params: { episodeID } }): Promise<res.IEpisodeComment[]> => {
      const ep = await fetcher.fetchSlimEpisodeByID(episodeID);
      if (!ep) {
        throw new NotFoundError(`episode ${episodeID}`);
      }
      const data = await db
        .select()
        .from(schema.chiiEpComments)
        .where(op.eq(schema.chiiEpComments.mid, episodeID));

      const userIDs = new Set(data.map((v) => v.uid));
      const users = await fetcher.fetchSlimUsersByIDs([...userIDs]);

      const comments: res.IEpisodeComment[] = [];
      const replies: Record<number, res.IEpisodeCommentBase[]> = {};

      for (const d of data) {
        const u = users[d.uid];
        if (!u) {
          continue;
        }
        if (d.related === 0) {
          comments.push(convert.toEpisodeComment(d, u));
        } else {
          const rs = replies[d.related] ?? [];
          rs.push(convert.toEpisodeCommentBase(d, u));
          replies[d.related] = rs;
        }
      }
      for (const comment of comments) {
        comment.replies = replies[comment.id] ?? [];
      }

      return comments;
    },
  );

  app.get('/subjects/-/episode/:episodeID/comments', (req, reply) => {
    const params = req.params as Record<string, string>;
    const episodeID = params.episodeID ?? '';
    return reply.redirect(`/p1/subjects/-/episodes/${episodeID}/comments`, 307);
  });

  app.post(
    '/subjects/-/episodes/:episodeID/comments',
    {
      schema: {
        summary: '创建条目的剧集吐槽',
        operationId: 'createSubjectEpComment',
        params: t.Object({
          episodeID: t.Integer({ examples: [1075440] }),
        }),
        tags: [Tag.Episode],
        response: {
          200: t.Object({
            id: t.Integer({ description: 'new reply id' }),
          }),
        },
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        body: req.Ref(req.CreateEpisodeComment),
      },
      preHandler: [requireLogin('creating a comment')],
    },
    /**
     * @param auth -
     * @param content - 吐槽内容
     * @param relatedID - 子吐槽的父吐槽ID，默认为 `0` 代表发送顶层吐槽
     * @param episodeID - 剧集 ID
     */
    async ({
      auth,
      body: { 'cf-turnstile-response': cfCaptchaResponse, content, replyTo = 0 },
      params: { episodeID },
    }) => {
      if (!(await turnstile.verify(cfCaptchaResponse))) {
        throw new CaptchaError();
      }
      if (!Dam.allCharacterPrintable(content)) {
        throw new BadRequestError('text contains invalid invisible character');
      }
      if (auth.permission.ban_post) {
        throw new NotAllowedError('create comment');
      }

      const ep = await fetcher.fetchSlimEpisodeByID(episodeID);
      if (!ep) {
        throw new NotFoundError(`episode ${episodeID}`);
      }

      if (replyTo !== 0) {
        const [parent] = await db
          .select({ id: schema.chiiEpComments.id, state: schema.chiiEpComments.state })
          .from(schema.chiiEpComments)
          .where(op.eq(schema.chiiEpComments.id, replyTo));
        if (!parent) {
          throw new NotFoundError(`parent comment id ${replyTo}`);
        }
        if (parent.state !== CommentState.Normal) {
          throw new NotAllowedError(`reply to a abnormal state comment`);
        }
      }

      await rateLimit(LimitAction.Subject, auth.userID);

      const reply: typeof schema.chiiEpComments.$inferInsert = {
        mid: episodeID,
        uid: auth.userID,
        related: replyTo,
        content: content,
        createdAt: DateTime.now().toUnixInteger(),
        state: CommentState.Normal,
      };
      const [result] = await db.insert(schema.chiiEpComments).values(reply);

      return { id: result.insertId };
    },
  );

  app.post('/subjects/-/episode/:episodeID/comments', (req, reply) => {
    const params = req.params as Record<string, string>;
    const episodeID = params.episodeID ?? '';
    return reply.redirect(`/p1/subjects/-/episodes/${episodeID}/comments`, 307);
  });

  app.put(
    '/subjects/-/episodes/-/comments/:commentID',
    {
      schema: {
        summary: '编辑条目的剧集吐槽',
        operationId: 'updateSubjectEpComment',
        params: t.Object({
          commentID: t.Integer({ examples: [1075440] }),
        }),
        tags: [Tag.Episode],
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
        body: req.Ref(req.UpdateEpisodeComment),
      },
      preHandler: [requireLogin('edit a comment')],
    },

    async ({ auth, body: { content }, params: { commentID } }) => {
      const [comment] = await db
        .select()
        .from(schema.chiiEpComments)
        .where(op.eq(schema.chiiEpComments.id, commentID));
      if (!comment) {
        throw new NotFoundError(`comment id ${commentID}`);
      }
      if (comment.uid !== auth.userID) {
        throw new NotAllowedError('edit a comment which is not yours');
      }
      if (comment.state !== CommentState.Normal) {
        throw new NotAllowedError(`edit to a abnormal state comment`);
      }

      const [{ replies = 0 } = {}] = await db
        .select({ replies: op.count() })
        .from(schema.chiiEpComments)
        .where(op.eq(schema.chiiEpComments.related, commentID));
      if (replies > 0) {
        throw new NotAllowedError('cannot edit a comment with replies');
      }

      await db
        .update(schema.chiiEpComments)
        .set({ content: content })
        .where(op.eq(schema.chiiEpComments.id, commentID));

      return {};
    },
  );

  app.put('/subjects/-/episode/-/comments/:commentID', (req, reply) => {
    const params = req.params as Record<string, string>;
    const commentID = params.commentID ?? '';
    return reply.redirect(`/p1/subjects/-/episodes/-/comments/${commentID}`, 307);
  });

  app.delete(
    '/subjects/-/episodes/-/comments/:commentID',
    {
      schema: {
        summary: '删除条目的剧集吐槽',
        operationId: 'deleteSubjectEpComment',
        params: t.Object({
          commentID: t.Integer({ examples: [1034989] }),
        }),
        tags: [Tag.Episode],
        security: [{ [Security.CookiesSession]: [], [Security.HTTPBearer]: [] }],
      },
      preHandler: [requireLogin('delete a comment')],
    },
    async ({ auth, params: { commentID } }) => {
      const [comment] = await db
        .select()
        .from(schema.chiiEpComments)
        .where(op.eq(schema.chiiEpComments.id, commentID));
      if (!comment) {
        throw new NotFoundError(`comment id ${commentID}`);
      }
      if (comment.uid !== auth.userID) {
        throw new NotAllowedError('delete a comment which is not yours');
      }
      if (comment.state !== CommentState.Normal) {
        throw new NotAllowedError('delete a abnormal state comment');
      }
      await db
        .update(schema.chiiEpComments)
        .set({ state: CommentState.UserDelete })
        .where(op.eq(schema.chiiEpComments.id, commentID));

      return {};
    },
  );

  app.delete('/subjects/-/episode/-/comments/:commentID', (req, reply) => {
    const params = req.params as Record<string, string>;
    const commentID = params.commentID ?? '';
    return reply.redirect(`/p1/subjects/-/episodes/-/comments/${commentID}`, 307);
  });
}
