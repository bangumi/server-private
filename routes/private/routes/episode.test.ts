import { beforeEach, describe, expect, test, vi } from 'vitest';

import { db, op } from '@app/drizzle/db.ts';
import * as schema from '@app/drizzle/schema';
import { emptyAuth } from '@app/lib/auth/index.ts';
import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './episode.ts';

beforeEach(async () => {
  await db.delete(schema.chiiEpComments).where(op.eq(schema.chiiEpComments.mid, 1027));
  await db.insert(schema.chiiEpComments).values({
    id: 205402,
    mid: 1027,
    content: '設計師太紳士了(bgm38)',
    state: 0,
    createdAt: 1400517144,
    uid: 382951,
    related: 0,
  });
  await db.insert(schema.chiiEpComments).values({
    id: 242518,
    mid: 1027,
    content: '结果现在再没有类似设定的了(?)。',
    state: 0,
    createdAt: 1413713418,
    uid: 382951,
    related: 205402,
  });
  await db.insert(schema.chiiEpComments).values({
    id: 215256,
    mid: 1027,
    content: 'op神曲~',
    state: 0,
    createdAt: 1404574453,
    uid: 382951,
    related: 0,
  });
});

describe('get episode', () => {
  test('should get episode', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/episodes/1027',
    });
    expect(res.json()).toMatchSnapshot();
  });
});

describe('get ep comment', () => {
  test('should get ep comment', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/episodes/1027/comments',
    });
    expect(res.json()).toMatchSnapshot();
  });
});

describe('create ep comment', () => {
  test('should create ep comment', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 382951,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: '/episodes/1027/comments',
      method: 'post',
      payload: { content: '114514', turnstileToken: 'fake-response' },
    });
    expect(res.statusCode).toBe(200);
    const pstID: number = res.json().id;
    const [pst] = await db
      .select()
      .from(schema.chiiEpComments)
      .where(op.eq(schema.chiiEpComments.id, pstID));
    expect(pst?.content).toBe('114514');
  });

  test('should now allowd create ep comment', async () => {
    const app = createTestServer();
    await app.register(setup);

    const res = await app.inject({
      url: '/episodes/1027/comments',
      method: 'post',
      payload: { content: '114514', turnstileToken: 'fake-response' },
    });
    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('edit ep comment', () => {
  test('should edit ep comment', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 382951,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/episodes/-/comments/215256`,
      method: 'put',
      payload: { content: 'new comment' },
    });
    expect(res.statusCode).toBe(200);

    const [data] = await db
      .select()
      .from(schema.chiiEpComments)
      .where(op.eq(schema.chiiEpComments.id, 215256));
    expect(data?.content).toBe('new comment');
  });

  test('should not edit ep comment not owned', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 382951 + 1,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/episodes/-/comments/205402`,
      method: 'put',
      payload: { content: 'new comment again' },
    });

    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });

  test('should not edit ep comment with replies', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 382951,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/episodes/-/comments/205402`,
      method: 'put',
      payload: { content: 'new comment again' },
    });

    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('delete ep comment', () => {
  test('not allowed not login', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      url: '/episodes/-/comments/114514',
      method: 'delete',
    });

    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });

  test('not allowed wrong user', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 1122,
      },
    });
    await app.register(setup);
    const res = await app.inject({
      url: '/episodes/-/comments/205402',
      method: 'delete',
    });

    expect(res.json()).toMatchSnapshot();
    expect(res.statusCode).toBe(401);
  });

  test('ok', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 382951,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      method: 'delete',
      url: '/episodes/-/comments/205402',
    });
    expect(res.json()).toMatchSnapshot();
    expect(res.statusCode).toBe(200);
    const [pst] = await db
      .select()
      .from(schema.chiiEpComments)
      .where(op.eq(schema.chiiEpComments.id, 205402));
    expect(pst?.state).toBe(6);
  });
});
