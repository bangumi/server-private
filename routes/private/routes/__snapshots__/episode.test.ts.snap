// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create ep comment > should now allowd create ep comment 1`] = `
Object {
  "code": "NEED_LOGIN",
  "error": "Unauthorized",
  "message": "you need to login before creating a comment",
  "statusCode": 401,
}
`;

exports[`delete ep comment > not allowed not login 1`] = `
Object {
  "code": "NEED_LOGIN",
  "error": "Unauthorized",
  "message": "you need to login before delete a comment",
  "statusCode": 401,
}
`;

exports[`delete ep comment > not allowed wrong user 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Unauthorized",
  "message": "you don't have permission to delete a comment which is not yours",
  "statusCode": 401,
}
`;

exports[`delete ep comment > ok 1`] = `Object {}`;

exports[`edit ep comment > should not edit ep comment not owned 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Unauthorized",
  "message": "you don't have permission to edit a comment which is not yours",
  "statusCode": 401,
}
`;

exports[`edit ep comment > should not edit ep comment with replies 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Unauthorized",
  "message": "you don't have permission to cannot edit a comment with replies",
  "statusCode": 401,
}
`;

exports[`get ep comment > should get ep comment 1`] = `
Array [
  Object {
    "content": "設計師太紳士了(bgm38)",
    "createdAt": 1400517144,
    "creatorID": 382951,
    "epID": 1027,
    "id": 205402,
    "reactions": Array [],
    "relatedID": 0,
    "replies": Array [
      Object {
        "content": "结果现在再没有类似设定的了(?)。",
        "createdAt": 1413713418,
        "creatorID": 382951,
        "epID": 1027,
        "id": 242518,
        "reactions": Array [],
        "relatedID": 205402,
        "state": 0,
        "user": Object {
          "avatar": Object {
            "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
            "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
            "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
          },
          "id": 382951,
          "joinedAt": 1512262276,
          "nickname": "树洞酱",
          "sign": "treeholechan@gmail.com 密码:lovemeplease",
          "username": "382951",
        },
      },
    ],
    "state": 0,
    "user": Object {
      "avatar": Object {
        "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
        "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
        "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
      },
      "id": 382951,
      "joinedAt": 1512262276,
      "nickname": "树洞酱",
      "sign": "treeholechan@gmail.com 密码:lovemeplease",
      "username": "382951",
    },
  },
  Object {
    "content": "op神曲~",
    "createdAt": 1404574453,
    "creatorID": 382951,
    "epID": 1027,
    "id": 215256,
    "reactions": Array [],
    "relatedID": 0,
    "replies": Array [],
    "state": 0,
    "user": Object {
      "avatar": Object {
        "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
        "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
        "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
      },
      "id": 382951,
      "joinedAt": 1512262276,
      "nickname": "树洞酱",
      "sign": "treeholechan@gmail.com 密码:lovemeplease",
      "username": "382951",
    },
  },
]
`;

exports[`get episode > should get episode 1`] = `
Object {
  "airdate": "2002-04-03",
  "comment": 16,
  "desc": "",
  "disc": 0,
  "duration": "24m",
  "id": 1027,
  "name": "ちぃ 目覚める",
  "nameCN": "叽，觉醒了",
  "sort": 1,
  "subject": Object {
    "id": 12,
    "images": Object {
      "common": "https://lain.bgm.tv/pic/cover/c/c2/0a/12_aDoa8.jpg",
      "grid": "https://lain.bgm.tv/pic/cover/g/c2/0a/12_aDoa8.jpg",
      "large": "https://lain.bgm.tv/pic/cover/l/c2/0a/12_aDoa8.jpg",
      "medium": "https://lain.bgm.tv/pic/cover/m/c2/0a/12_aDoa8.jpg",
      "small": "https://lain.bgm.tv/pic/cover/s/c2/0a/12_aDoa8.jpg",
    },
    "info": "27话 / 2002年4月2日 / 浅香守生 / CLAMP / 阿部恒",
    "locked": false,
    "name": "ちょびっツ",
    "nameCN": "人形电脑天使心",
    "nsfw": false,
    "type": 2,
  },
  "subjectID": 12,
  "type": 0,
}
`;
